File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # 自底向上返回节点高度（子树高度信息）
        # -999 作为不平衡树的信号标记
        def dfs(node):
            if node is None:
                return 0

            left_tree = dfs(node.left)
            if left_tree == -999:  # 不平衡树提前剪枝
                return -999
            right_tree = dfs(node.right)
            if right_tree == -999: # 不平衡树提前剪枝
                return -999
            if abs(left_tree - right_tree) > 1:
                return -999

            return 1 + max(left_tree, right_tree)

        return dfs(root) != -999

            

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean result = true;
    public boolean isBalanced(TreeNode root) {
        maxDepth(root);
        return result;
    }
    int maxDepth(TreeNode node) {
        // base case
        if(node == null) return 0;

        int leftMaxDepth = maxDepth(node.left);
        int rightMaxDepth = maxDepth(node.right);

        // 后序遍历操作位置，继续传递参数
        if(Math.abs(leftMaxDepth - rightMaxDepth) > 1) {
            result = false;
        }
        return 1 + Math.max(leftMaxDepth, rightMaxDepth);
    }
}